src/lib/registration/addon.rb

```diff
@@ -84,25 +83,70 @@ def registered_not_installed
 
       # create an Addon from a SUSE::Connect::Product
       # @param root [SUSE::Connect::Product] the root add-on object
-      # @return [Addon] created Addon object
+      # @return [Array<Addon>] list of addons, where the first one is
+      #   the one based on root and rest is its children
       def create_addon_with_deps(root)
-        root_addon = Addon.new(root)
-        result = [root_addon]
-
-        (root.extensions || []).each do |ext|
-          child = create_addon_with_deps(ext)
-          result.concat(child)
-          child.first.depends_on = root_addon
-          root_addon.children << child.first
+        # to_process is array of pairs, where first is pure addon to process and second is
+        # its dependency. Currently SUSE::Connect structure have only one dependency.
+        to_process = [[root, nil]]
+        processed = Set.new
+        result = []
+
+        to_process.each do |(pure, dependency)|
+          next if processed.include?(pure)
```

Comment:
NP: I'd add a comment saying "avoid infinite loop when there is a cyclic dependency", just you can quickly see that this case is also handled here.